src/Controller/RepeatScheduleController.php

```diff
@@ -0,0 +1,425 @@
+ *   enabling real-time updates to scheduling views without requiring page
+ *   reloads.
+ *
+ * Methods such as `getDateRangeData`, `processResult`, `joinTables`, and
+ * others work together to query, aggregate, and format the necessary data from
+ * the database. This controller thus serves as a crucial component in
+ * applications requiring detailed and flexible handling of repeating events and
+ * schedules.
+ */
+class RepeatScheduleController extends RepeatController {
+
+  /**
+   * The config factory service.
+   *
+   * @var \Drupal\Core\Config\ConfigFactoryInterface
+   */
+  protected $configFactory;
+
+  /**
+   * {@inheritdoc}
+   */
+  public static function create(ContainerInterface $container): self {
+    $instance = parent::create($container);
+    $instance->configFactory = $container->get('config.factory');
+    return $instance;
+  }
+
+  /**
+   * Retrieves event data within a specified date range, filtered by location
+   * and categories.
+   *
+   * This method constructs a database query to fetch data for repeat events
+   * from the 'repeat_event' table. It dynamically joins additional tables,
+   * adds necessary fields, and applies filtering conditions based on
+   * the input parameters: start date, end date, categories, and location.
+   * The method is designed to provide comprehensive event data suitable for
+   * rendering events in calendar views or for other scheduling purposes.
+   *
+   * The query execution results are then processed to format the data
+   * appropriately for the consuming client, adjusting the date to the
+   * beginning of the week if necessary to align with calendar week views.
+   *
+   * @param Request $request
+   *   The current request object, potentially containing additional parameters
+   *   or filters to apply to the query.
+   * @param string $location
+   *   The location identifier to filter the events by.
+   * @param string $start_date
+   *   The start date of the date range.
+   * @param string $end_date
+   *   The end date of the date range.
+   * @param string $categories
+   *   A comma-separated list of category identifiers to filter the events.
+   *   This allows for retrieval of events that fall into specific categories.
+   *
+   * @return array
+   *   An array of processed event data, formatted for easy consumption by the
+   *   caller. The structure of this array and the specific fields included may
+   *   depend on the implementation of `processResult` and the needs of the
+   *   consuming client or view.
+   */
+  public function getDateRangeData(Request $request, string $location, string $start_date, string $end_date, string $categories): array {
+    $query = $this->database->select('repeat_event', 're');
+    $this->joinTables($query);
+    $this->addFields($query);
+    $this->addCondition($query, $start_date, $end_date, $categories, $location, $request);
+    $result = $query->execute()->fetchAll();
```

Comment:
can we have a limit of maximum requested from DB?
For security and performance reason

Imagine you have 5+ years of data

Reply:
This request to the database is limited to the current week of viewing, that is, we can get data for only one week, and there is always one request for the user in the database. Caching may be added later.

Reply:
We can add limit to this query as well, but in this case we do not guarantee that the user will see all events